test/unit/common/test_shardtrie.py

```diff
@@ -0,0 +1,312 @@
+        self.assertEqual(
+            len(list(trie.get_important_nodes(limit = 2))),
+            2)
+
+        # Iterate starting after node abc
+        marker = trie.get_node('abc')
+        self.assertEqual(
+            list(trie.get_important_nodes(marker=marker))[0].full_key(),
+            'abd')
+
+        # Iterate starting at 'abc'
+        self.assertEqual(
+            list(trie.get_important_nodes('abc'))[0].full_key(),
+            'abc')
+
+    def test_shardtrie_dump(self):
+        """Test adding a node to a distributed shardtrie"""
+        trie = self.trie
+        self.assertEqual(trie.dump(), json.loads(self._test_shard_trie))
+
+    def test_shardtrie_get_last_node(self):
+        """Test getting the last node in the shardtrie"""
+        trie = self.trie
+        last_node = trie.get_last_node()
+        self.assertEqual(last_node.key, 'd')
+
+    def test_shardtrie_get_distributed_nodes(self):
+        """Test getting distributed nodes after shard"""
+        trie = self.trie
+
+        distributed_nodes = trie.get_distributed_nodes()
+        self.assertEqual(len(list(distributed_nodes)), 0)
+
+        subtrie = trie.split_trie('ab')
+
+        distributed_nodes = trie.get_distributed_nodes()
+        self.assertEqual(len(list(distributed_nodes)), 1)
+
+    def test_shardtrie_get_large_subtries(self):
+        """Test getting subtries over certain size"""
+        trie = self.trie
+
+        for i in range(20):
+            trie.add('abc%d' % i)
+            trie.add('abd%d' % i)
+
+        large_tries = [x[2] for x in trie.get_large_subtries(15)]
+        self.assertEqual(large_tries, ['abd', 'abc', 'ab'])
+
+    def test_shardtrie_trim_trunk(self):
+        """Test trimming the trunk of the shard trie"""
+        trie = self.trie
+
+        # Call on an already trimmed trie
+        trie.trim_trunk()
+
+        # Call on an trimmed trie
+        subtrie = trie.split_trie('ab')
+        subtrie.trim_trunk()
+
+        # Call on an untrimmed trie
+        subtrie.add('abef', flag=shardtrie.EMPTY)
+        subtrie.add('abefg', 'test data')
+        subtrie.trim_trunk()
+
+class TestCountingTrie(unittest.TestCase):
+    """Tests for swift.common.CountingTrie"""
+
+    def setUp(self):
+        pass
+
+    def tearDown(self):
+        pass
+
+    def _setup_trie(self, keys=['a', 'abc', 'abd'], group_count=500):
+        trie = shardtrie.CountingTrie(max_group_size=group_count)
+
+        for node in keys:
+            trie.add(node)
+
+        return trie
+
+    def test_countingtrie_add_distributed(self):
+        trie = self._setup_trie(group_count=3)
+
+        trie.add('aa', True)
+        self.assertTrue(trie._root._children['a']._children['a']._distributed)
+
+    def test_countingtrie_get_best_candidate(self):
+        """Test getting the best candidate found by the counting trie"""
+        trie = self._setup_trie(group_count=2)
+
+        self.assertEqual(trie.get_best_candidates(), (2, ['ab']))
```

Comment:
There is also a case where the counting trie may find more then 1 candidate at the lowest level in the trie, so we should test that too, maybe after this assert (line 279) we add:

  keys = ['a', 'abc', 'abd', 'cba', 'cbb']
  trie = self._setup_trie(keys=keys, group_count=2)
  self.assertEqual(trie.get_best_candidates(), (2, ['ab', 'cb']))